kedro/contrib/io/matplotlib/matplotlib_s3_writer.py

```diff
@@ -0,0 +1,173 @@
+            filepath: Path to a matplot object file.
+            boto_session_args: See
+                https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html
+            s3_client_args: See
+                https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#client
+            s3_put_object_args: See
+                https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.put_object
+            credentials: A dictionary of s3 access and secret keys.
+                Must contain ``aws_access_key_id`` and ``aws_secret_access_key``.
+                Updates ``s3_client_args`` if provided.
+            load_args: Currently ignored as loading is not supported.
+            save_args: Save args passed to `plt.savefig`. See
+                https://matplotlib.org/api/_as_gen/matplotlib.pyplot.savefig.html
+        """
+
+        self._boto_session_args = boto_session_args or {}
+        self._s3_client_args = s3_client_args or {}
+        self._credentials = copy.deepcopy(credentials) or {}
+        self._s3_put_object_args = s3_put_object_args or {}
+
+        if self._credentials:
+            self._s3_client_args["aws_access_key_id"] = self._credentials[
+                "aws_access_key_id"
+            ]
+            self._s3_client_args["aws_secret_access_key"] = self._credentials[
+                "aws_secret_access_key"
+            ]
+
+        self._filepath = filepath
+        self._load_args = load_args if load_args else dict()
+        self._save_args = save_args if save_args else dict()
+        self._bucket = bucket
+
+    def _describe(self) -> Dict[str, Any]:
+        return dict(
+            bucket=self._bucket, load_args=self._load_args, save_args=self._save_args
+        )
+
+    def _load(self) -> None:
+        raise DataSetError(
+            "Loading not supported for `{}`".format(self.__class__.__name__)
+        )
+
+    def _save(self, data: Union[figure, List[figure], Dict[str, figure]]) -> None:
+        if isinstance(data, list):
+            for index, plot in enumerate(data):
+                key_path = self._filepath + "/" + str(index) + ".png"
+                self._save_to_s3(key_name=key_path, plot=plot)
+
+        elif isinstance(data, dict):
+            for plot_name, plot in data.items():
+                key_path = self._filepath + "/" + plot_name
+                self._save_to_s3(key_name=key_path, plot=plot)
+
+        else:
+            self._save_to_s3(key_name=self._filepath, plot=data)
+
+    def _save_to_s3(self, key_name, plot):
+
+        bytes_object = io.BytesIO()
+        plot.savefig(bytes_object, **self._save_args)
+
+        session = boto3.Session(**self._boto_session_args)
+
+        session.client("s3", **self._s3_client_args).put_object(
+            Bucket=self._bucket,
+            Key=key_name,
+            **self._s3_put_object_args,
+            Body=bytes_object.getvalue(),
+        )
```

Comment:
The way I read this, it create a new boto client and a new boto session for every figure we save from the list/dict. Is there a specific reason for this, or can we just move these to live on the class (i.e. in the `__init__`) instead?

Reply:
We've also moved all our S3 datasets to use `s3fs` under the hood - would be ideal if we could keep it consistent across the board. This PR's approach customises a lot more kwargs than we do currently by default (see for instance `CSVS3DataSet`), I wonder what's the motivation behind it?
If they're all definitely required, it seems like all the different types of kwargs could be passed in on instantiating an `S3FileSystem`:
`s3_client_args` -> `client_kwargs`
`boto_session_args` -> `kwargs`
`s3_put_object_args` -> could perhaps be rolled into `s3_additional_kwargs`

Reply:
Regarding comment 1: 
No specific reason, I was just trying to put all the s3 related stuff under 1 function, so that if someone else needs to modify for Azure or GCP (or who knows what else), the only modification they need to do is the `_save_to_s3` method and possibly the line above where `_save_to_s3` is called (for the default filenames). 

For comment 2:
Yes I noticed that it was different. Definitely agree to make things consistent. Let me try. I hacked this quite quickly together on a project using `boto` instead of `s3fs` hence why the PR uses the package.

Could you clarify on "customises lots of defaults"? There's only `_credentials`. Do you mean why so many options for args? ie `s3_client_args`, `boto_session_args`, `s3_put_object_args`? On occasion, this level of kwargs are required. You'll see there's a specific test for encryption because it's something I faced on client environment. On NC1, we don't face this level of security/issues, our bucket policy is very relaxed. So definitely a way is needed to pass all other possible arguments into `S3FileSystem`.